torch/csrc/autograd/function_port.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+
+namespace torch {
+namespace autograd {
+
+class Function;
+class Variable;
+
+/// Represents an input or "port" of a function.
+struct FunctionPort {
```

Comment:
I like Edge as name for this as long as we are explicit in the comments, and consistent with the names of fields:

```
// An edge (func, index) represents the index-th input of `func`
// Functions contain a list `output_edges` the same length as the number of outputs it produces which connects output to a particular input of another function.
// In this representation a particular function input (func, index) may have multiple incoming edges, which are implicitly summed before passing the value to func.
```

I think we should avoid using `output_nr` in Edge because the offset is the index into the _input_ list of the function.  `output_nr` can also be confusing in Variables, because depending on how the actual gradient Functions were produced, it doesn't actually correspond to the position in the output list a Variable was in (e.g. if a single forward produced 2 separate backward Functions for its two inputs).

Reply:
Yes, edges hold only an **input_nr**, never an output_nr. It's just that when you take `gradient_edge()` outputs become reversed with inputs, so we set the `input_nr` of the edge to equal to the `output_nr` of the Variable.

Reply:
Is that always true?

```
  def f(c,d):  return c*2, c*3
  a, b = f(c);
```

In this case, someone might implement it in autograd two separate gradient functions:
```
def f'1(da) return 2*da
def f'1(db) return 3*db 
a.grad_fn = f'1
a.output_nr = 0
b.grad_fn = f'2
b.output_nr = 0
```
In this case, `output_nr` is not an accurate name since `b` was never the 0-output of anything.

Reply:
Yeah that is theoretically allowed, but we never do (nor did) that. `grad_input_nr` would be more precise, but let's limit the number of codemods we'll be running. This one makes sense, because the API will be more convenient and it will be used in external modules, but in general you probably won't be using `output_nr()` outside of autograd.

Reply:
Sure -- I expect our usage of output_nr will be limited once we flip most things over to `gradient_edge()` anyway.